### PR TITLE
format the header of files

### PR DESCRIPTION
**Pull request by yhmtsai**
This PR sorts the ordering of headers, use correct `"header"` or `<header>`, and correct the definition of header.

1. Add `hip/hip_runtime.h` into `hip/base/config.hip.hpp` as its main include to ensure it is in front of `hip/device_functions.h` mentioned in #412 
2. use `IncludeBlocks: Regroup` (need clang-format 6) to sort the header except the main include
3. add config (`dev_tools/scripts/config`) to set the rules how to find the corresponding main include
4. if there are something like comment of #if block before `#include`, the script will give alarm "sorting is maybe not correct"
5. Also alarm if there are multiple matches the main include.

Also mention this in #389 , but remove many if-else and unclear regex (which is almost handled by config)

add target `format_header` and `format_header_all` in CMake
`format_header` only process the changed files of branch
`format_header_all` process all code files.
They ignore build, examples, third_party, ginkgo.hpp

add `// force-top: on` and `// force-top: off` to keep some headers in the top (also before the main header).
I change the warning condition. If there are comments or #if block before codes, it gives warning.
The following is warning
```
Warning benchmark/utils/formats.hpp: sorting maybe is not correct
Warning core/base/extended_float.hpp: sorting maybe is not correct
Warning core/device_hooks/common_kernels.inc.cpp: sorting maybe is not correct
Warning include/ginkgo/core/base/abstract_factory.hpp: sorting maybe is not correct
Warning include/ginkgo/core/base/name_demangling.hpp: sorting maybe is not correct
Warning include/ginkgo/core/base/types.hpp: sorting maybe is not correct
Warning include/ginkgo/core/base/utils.hpp: sorting maybe is not correct
Warning reference/test/stop/combined.cpp: sorting maybe is not correct
Warning reference/test/stop/time.cpp: sorting maybe is not correct
```

### CODE REVIEW

**Comment by yhmtsai** [comment]
6 parts for reviews
- config/scripts [review](https://github.com/ginkgo-project/ginkgo/pull/447/files/55a80da0a4247401db21d5ae4932930a3d01b176)
- core changes [review](https://github.com/ginkgo-project/ginkgo/pull/447/commits/1290b812da4e4764f7f6757d33a63279a5150c56)
- cuda changes [review](https://github.com/ginkgo-project/ginkgo/pull/447/commits/1e5247cb6e4944498172938a413116eb55d65608)
- hip changes [review](https://github.com/ginkgo-project/ginkgo/pull/447/commits/6f7f948052b87797e0e0b7f77992ccdc57a79d75)
- include/ginkgo/core changes [review](https://github.com/ginkgo-project/ginkgo/pull/447/commits/908f50ed5299674f838b69c9fa0beebddaaa5b7e)
- omp changes [review](https://github.com/ginkgo-project/ginkgo/pull/447/commits/2f019d3dfebb1dbeec3a336ba1efbbe0f621db97)
- referece changes [review](https://github.com/ginkgo-project/ginkgo/pull/447/commits/3786f9470cb90d1bfa11b90a1550d60117aeae43)

Also, the following is the alarm of script
```bash
(base) mike:ginkgo$ find core -type f -not -name "*.txt" -exec  dev_tools/scripts/format_header.sh {} \;
(base) mike:ginkgo$ find cuda -type f -not -name "*.txt" -exec  dev_tools/scripts/format_header.sh {} \;
(base) mike:ginkgo$ find hip -type f -not -name "*.txt" -exec  dev_tools/scripts/format_header.sh {} \;
(base) mike:ginkgo$ find include/ginkgo/core -type f -not -name "*.txt" -exec  dev_tools/scripts/format_header.sh {} \;
ALARM include/ginkgo/core/base/name_demangling.hpp may not be sorted correctly
(base) mike:ginkgo$ find omp -type f -not -name "*.txt" -exec  dev_tools/scripts/format_header.sh {} \;
(base) mike:ginkgo$ find reference -type f -not -name "*.txt" -exec  dev_tools/scripts/format_header.sh {} \;
```
In [name_demangling.hpp](https://github.com/ginkgo-project/ginkgo/blob/3786f9470cb90d1bfa11b90a1550d60117aeae43/include/ginkgo/core/base/name_demangling.hpp#L37-L45), it contains `#if` block before other `#include`

However, if #if block is in the end, it will not give an alarm like [reference/test/stop/combined.cpp](https://github.com/ginkgo-project/ginkgo/blob/3786f9470cb90d1bfa11b90a1550d60117aeae43/reference/test/stop/combined.cpp#L43-L47) and [reference/test/stop/time.cpp](https://github.com/ginkgo-project/ginkgo/blob/3786f9470cb90d1bfa11b90a1550d60117aeae43/reference/test/stop/time.cpp#L40-L43)

File: dev_tools/scripts/format_header.sh [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+get_include_regex () {
+    local file="$1"
+    local core_suffix=""
+    local path_prefix=""
+    local path_ignore="0"
+    local fix_include=""
+    local remove_test="false"
+    local item_regex="^-\ +\"(.*)\""
+    local path_prefix_regex="PathPrefix:\ +\"(.*)\""
```
Comment (upsj) [inline dev_tools/scripts/format_header.sh]: Can you maybe add a comment explaining the effect of these configuration options? Some are obvious, for others you might have to look closer into the code.

File: cuda/components/prefix_sum.cu [hunk 2]
```diff
@@ -30,10 +30,10 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************<GINKGO LICENSE>*******************************/
 
-#include "core/components/prefix_sum.hpp"
+#include "cuda/components/prefix_sum.cuh"
 
 
-#include "cuda/components/prefix_sum.cuh"
+#include "core/components/prefix_sum.hpp"
```
Comment (upsj) [inline cuda/components/prefix_sum.cu]: I would rather leave the previous order, as this is the implementation of the prefix_sum kernel declared in `core/components/prefix_sum.hpp`. Does this have to do with the fact that the file name does not end with `kernels`?

File: cuda/test/components/prefix_sum.cu [hunk 3]
```diff
@@ -44,6 +41,9 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <ginkgo/core/base/array.hpp>
 
 
+#include "core/components/prefix_sum.hpp"
```
Comment (upsj) [inline cuda/test/components/prefix_sum.cu]: I would prefer to leave this header at the top since it is the component under test

File: hip/components/prefix_sum.hip.cpp [hunk 4]
```diff
@@ -30,10 +30,10 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************<GINKGO LICENSE>*******************************/
 
-#include "core/components/prefix_sum.hpp"
+#include "hip/components/prefix_sum.hip.hpp"
 
 
-#include "hip/components/prefix_sum.hip.hpp"
+#include "core/components/prefix_sum.hpp"
```
Comment (upsj) [inline hip/components/prefix_sum.hip.cpp]: Same here

File: hip/test/components/prefix_sum.hip.cpp [hunk 5]
```diff
@@ -44,6 +41,9 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <ginkgo/core/base/array.hpp>
 
 
+#include "core/components/prefix_sum.hpp"
```
Comment (upsj) [inline hip/test/components/prefix_sum.hip.cpp]: Same here

File: omp/test/components/prefix_sum.cpp [hunk 6]
```diff
@@ -44,6 +41,9 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <ginkgo/core/base/array.hpp>
 
 
+#include "core/components/prefix_sum.hpp"
```
Comment (upsj) [inline omp/test/components/prefix_sum.cpp]: Same here

File: reference/test/components/prefix_sum.cpp [hunk 7]
```diff
@@ -41,6 +38,9 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <gtest/gtest.h>
 
 
+#include "core/components/prefix_sum.hpp"
```
Comment (upsj) [inline reference/test/components/prefix_sum.cpp]: Same here

**Comment by yhmtsai** [comment]
@upsj How do you think about combining prefix_sum.hip.hpp and prefix_sum.hip.cpp into prefix_sum_kernels.hip.cpp?
1. the interface is already defined in core/components/prefix_sum.hpp
2. It will be the same as the others like matrix and solver (with kernels suffix and the main header in core)

**Comment by upsj** [comment]
> @upsj How do you think about combining prefix_sum.hip.hpp and prefix_sum.hip.cpp into prefix_sum_kernels.hip.cpp?
> 
>     1. the interface is already defined in core/components/prefix_sum.hpp
> 
>     2. It will be the same as the others like matrix and solver (with kernels suffix and the main header in core)

My answer to that would probably be "yes and/or unsure": At the moment, prefix_sum.hip.hpp only contains things that are used in prefix_sum.hip.cpp. For ParILUT however, I need a "wide" prefix sum kernel that is quite specific to the selection implementation and uses device functions I also put into prefix_sum.hip.hpp (or more specifically, prefix_sum.hpp.inc). So what do you think makes more sense:
* Leave header and source separated, but rename the source to _kernels?
* Combine header and source and later on also add the `wide_prefix_sum` to this file, even though it is only used by ParILUT?

File: core/device_hooks/reference_hooks.cpp [hunk 8]
```diff
@@ -30,7 +30,6 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************<GINKGO LICENSE>*******************************/
 
-
 #include <ginkgo/core/base/version.hpp>
```
Comment (pratikvn) [inline core/device_hooks/reference_hooks.cpp]: I could not find this rule of "single empty line after license beacon" . Maybe you can add it to the contribution guidelines ?

File: core/matrix/dense_kernels.hpp [hunk 9]
```diff
@@ -34,10 +34,12 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #define GKO_CORE_MATRIX_DENSE_KERNELS_HPP_
 
 
-#include <ginkgo/core/base/types.hpp>
 #include <ginkgo/core/matrix/dense.hpp>
 
 
+#include <ginkgo/core/base/types.hpp>
+
```
Comment (pratikvn) [inline core/matrix/dense_kernels.hpp]: The standard headers such as `<memory>` and possibly others are missing here. Is it possible to fix the proper standard headers here and other files as well ? Or would you prefer it be done in another PR ?

File: core/device_hooks/reference_hooks.cpp (hunk 8 above)
Comment (thoasm) [reply core/device_hooks/reference_hooks.cpp]: We actually don't have this rule and we have both versions (1 and 2 empty lines) present in our files. I think it is fine to allow and have both in our files (the contributing guidelines are quite loaded already).

File: core/test/matrix/coo_builder.cpp [hunk 10]
```diff
@@ -30,15 +30,15 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************<GINKGO LICENSE>*******************************/
 
-#include "core/matrix/coo_builder.hpp"
```
Comment (thoasm) [inline core/test/matrix/coo_builder.cpp]: I think this is the correct spot for this header.

File: core/test/matrix/csr_builder.cpp [hunk 11]
```diff
@@ -30,15 +30,15 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************<GINKGO LICENSE>*******************************/
 
-#include "core/matrix/csr_builder.hpp"
```
Comment (thoasm) [inline core/test/matrix/csr_builder.cpp]: Same here, I think this is / was the correct position for this include.

File: core/device_hooks/reference_hooks.cpp (hunk 8 above)
Comment (yhmtsai) [reply core/device_hooks/reference_hooks.cpp]: add_license.sh use single empty line after license.
I am fine with whether add the rule into guidelines or not, but I would like to keep the same behavior as the current script.

Comment (thoasm) [reply core/device_hooks/reference_hooks.cpp]: I am fine with having just one empty line (and that the script does it that way), I would just like to not add it to the guidelines because it is quite unimportant in my opinion, and I don't want to increase the guidelines with such a detail.

File: core/matrix/dense_kernels.hpp (hunk 9 above)
Comment (yhmtsai) [reply core/matrix/dense_kernels.hpp]: I think it is better in another PR (maybe including `include-what-you-use`).
For the `include-what-you-use`, add the recommdation in the begin and remove the unneeded header. Moreover, always gives warning to check it again.
If there is a #if-#else block, it is hard to insert them at the correct position.

**Comment by yhmtsai** [comment]
@upsj i go for the option 1 but change others to _kernels in the new commit. ( Maybe use _impl or something else as the keywords said the main header is in the same folder)
In components, if the filename contain _kernels, its main header is in the same folder.
If not, its main header is the header with the same name in core, which is like the behaviour of other files  in matrix or solver.

File: omp/test/solver/cg_kernels.cpp [hunk 12]
```diff
@@ -33,21 +33,23 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
 #include <ginkgo/core/base/executor.hpp>
 #include <ginkgo/core/matrix/dense.hpp>
 #include <ginkgo/core/stop/combined.hpp>
 #include <ginkgo/core/stop/iteration.hpp>
 #include <ginkgo/core/stop/residual_norm_reduction.hpp>
 
+
+#include "core/solver/cg_kernels.hpp"
+#include "core/test/utils.hpp"
+
```
Comment (thoasm) [inline omp/test/solver/cg_kernels.cpp]: Missing empty line.

File: omp/test/solver/fcg_kernels.cpp [hunk 13]
```diff
@@ -33,21 +33,23 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
 #include <ginkgo/core/base/executor.hpp>
 #include <ginkgo/core/matrix/dense.hpp>
 #include <ginkgo/core/stop/combined.hpp>
 #include <ginkgo/core/stop/iteration.hpp>
 #include <ginkgo/core/stop/residual_norm_reduction.hpp>
 
+
+#include "core/solver/fcg_kernels.hpp"
+#include "core/test/utils.hpp"
+
```
Comment (thoasm) [inline omp/test/solver/fcg_kernels.cpp]: Missing empty line.

File: reference/test/matrix/hybrid_kernels.cpp [hunk 14]
```diff
@@ -30,23 +30,23 @@ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************<GINKGO LICENSE>*******************************/
 
-#include "core/matrix/hybrid_kernels.hpp"
-
-
```
Comment (thoasm) [inline reference/test/matrix/hybrid_kernels.cpp]: I think this should stay since it is the header that is tested.

Comment (yhmtsai) [reply reference/test/matrix/hybrid_kernels.cpp]: For my original config, it should use <ginkgo/core/matrix/hybrid.hpp>.
I choose use <ginkgo/core/matrix/hybrid.hpp> not "core/matrix/hybrid_kernels.hpp" because it use apply to test and do not call the function directly.
If the script does not find the desired main header, it will ignore the main header part.

Comment (thoasm) [reply reference/test/matrix/hybrid_kernels.cpp]: It is fine with me if you add `ginkgo/core/matrix/hybrid.hpp` as the top include file.
I am not sure if we use any functions of that header, but to me, every test should have the corresponding header as the very first include file.

File: core/matrix/dense_kernels.hpp (hunk 9 above)
Comment (tcojean) [reply core/matrix/dense_kernels.hpp]: `include-what-you-use` is sometimes wrong, so I would not directly include its result as part of a script. Maybe it can be configured to always provide the correct solution, but this requires some work.

See this for example https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/jobs/404997031#L2851
`name_demangling` is required, we use in stream `name_demangling::get_dynamic_type`, but it says it should be removed...

**Comment by yhmtsai** [comment]
It is updated.

add target `format_header` and `format_header_all` in CMake
`format_header` only process the changed files of branch
`format_header` process all code files.
They ignore build, examples, third_party, ginkgo.hpp

add `// force-top: on` and `// force-top: off` to keep some headers in the top (also before the main header).
I change the warning condition. If there are comments or #if block before codes, it gives warning.
The following is warning
```
Warning benchmark/utils/formats.hpp: sorting maybe is not correct
Warning core/base/extended_float.hpp: sorting maybe is not correct
Warning core/device_hooks/common_kernels.inc.cpp: sorting maybe is not correct
Warning include/ginkgo/core/base/abstract_factory.hpp: sorting maybe is not correct
Warning include/ginkgo/core/base/name_demangling.hpp: sorting maybe is not correct
Warning include/ginkgo/core/base/types.hpp: sorting maybe is not correct
Warning include/ginkgo/core/base/utils.hpp: sorting maybe is not correct
Warning reference/test/stop/combined.cpp: sorting maybe is not correct
Warning reference/test/stop/time.cpp: sorting maybe is not correct
```

File: core/test/solver/gmres.cpp [hunk 15]
```diff
@@ -47,6 +46,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <ginkgo/core/stop/residual_norm_reduction.hpp>
 
 
+#include "core/solver/gmres.cpp"
```
Comment (pratikvn) [inline core/test/solver/gmres.cpp]: A `.cpp` file ?

File: cuda/test/base/exception_helpers.cu [hunk 16]
```diff
@@ -33,12 +33,10 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <ginkgo/core/base/exception_helpers.hpp>
 
 
-#include <gtest/gtest.h>
-
-
 #include <cublas_v2.h>
 #include <cuda_runtime.h>
 #include <cusparse.h>
+#include <gtest/gtest.h>
```
Comment (pratikvn) [inline cuda/test/base/exception_helpers.cu]: I understand that these are third party libraries and hence are bunched. But maybe it is better if `gtest/gtest.h` is separate ?

Comment (yhmtsai) [reply cuda/test/base/exception_helpers.cu]: If we need to separate them, I prefer to split this section to device-related library (cuda/hip/omp) and other third party libraries.
Otherwise, I would like to keep them together.

Comment (pratikvn) [reply cuda/test/base/exception_helpers.cu]: Yes, device related into one and third party into one would be good.

File: dev_tools/scripts/format_header.sh (hunk 1 above)
Comment (yhmtsai) [reply dev_tools/scripts/format_header.sh]: Do you mean this comment?
I write some comments to explain the configuration option [here](https://github.com/ginkgo-project/ginkgo/pull/447/files#diff-e7e36e51d1a12bf3793383440bca985cR45-R61)
Or, do I misunderstand what you mean?

Comment (upsj) [reply dev_tools/scripts/format_header.sh]: Thanks, that was what I was looking for!

**Comment by pratikvn** [comment]
Just a question, is this a blocking change ? I see that if `GINKGO_DEVEL_TOOLS` is `on`, then the `format_header` script is called. Currently, the user needs to call `make  format_header_all` to format the files. Do we want to do something similar as for `make format` ? Or is the way it is now, enough ?

File: dev_tools/scripts/format_header.sh [hunk 17]
```diff
@@ -0,0 +1,313 @@
...
+# It reads "dev_tools/scripts/config" to generate the corresponding main header
+# The setting setting:
+# - "file_regex"
+#   - CoreSuffix: "core_suffix_regex"           (default "")
+#   - PathPrefix: "path_prefix_regex"           (default "")
+#   - PathIgnore: "path_ignore_number"          (default "0")
+#   - RemoveTest: "false/true"                  (default "test")
+#   - FixInclude: "the specific main header"    (default "")
+# Only "file_regex" without any setting is fine, and it means find the same name with header suffix
+# For exmaple, /path/to/file.cpp will /path/to/file.hpp
```
Comment (pratikvn) [inline dev_tools/scripts/format_header.sh:54]:
```suggestion
# For example, /path/to/file.cpp will change to /path/to/file.hpp
```

File: dev_tools/scripts/format_header.sh [hunk 18]
```diff
@@ -0,0 +1,313 @@
...
+            if [[ "${line}" =~ $END_BLOCK_REX ]]; then
+                IN_BLOCK=$((IN_BLOCK-1))
+            fi
+        else
+            echo "${line}" >> ${BEFORE}
+        fi
+    fi
+done < $1
+if [ "${ALARM}" = "true" ]; then
+    echo "Warning $1: sorting maybe is not correct"
```
Comment (pratikvn) [inline dev_tools/scripts/format_header.sh:214]:
```suggestion
    echo "Warning $1: sorting is probably incorrect"
```

File: dev_tools/scripts/format_header.sh [hunk 19]
```diff
@@ -0,0 +1,313 @@
...
+    echo "" >> $1
+    rm "${FORCE_TOP}"
+fi
+
+# Write the main header and give warnning if there are multiple matches
+if [ -f "${BEFORE}" ]; then
+    # sort or remove the duplication
+    clang-format -i -style=file ${BEFORE}
+    if [ $(wc -l < ${BEFORE}) -gt "1" ]; then
+        echo "Warning $1: there are multiple main header matched"
```
Comment (pratikvn) [inline dev_tools/scripts/format_header.sh:254]:
```suggestion
        echo "Warning $1: there are multiple main header matchings"
```

File: dev_tools/scripts/format_header.sh [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+        echo "#include <hip/hip_runtime.h>" > temp
+    fi
+    head -n -${KEEP_LINES} ${CONTENT} >> temp
+    if [ ! -z "${IFNDEF}" ] && [ ! -z "${DEFINE}" ]; then
+        # Ignore the last line #endif
+        if [[ "${LAST_NONEMPTY}" =~ $ENDIF_REX ]]; then
+            head -n -1 temp > ${CONTENT}
+            echo "#endif  // $HEADER_DEF" >> ${CONTENT}
+        else 
+            echo "Warning $1: Found the begin header_def but do not find the end of header_def"
```
Comment (pratikvn) [inline dev_tools/scripts/format_header.sh:277]:
```suggestion
            echo "Warning $1: Found the begin header_def but did not find the end of header_def"
```

**Comment by yhmtsai** [comment]
@pratikvn No, it is not a blocking changes now.
Currently, I would prefer to leave it as an option.
we can try it for a while. If everything is expected, we can also add it like git-cmake-format.
